app/src/animation/animation.c

```diff
@@ -0,0 +1,167 @@
+
+/**
+ * Size of the animation device pointers array.
+ */
+static const size_t animations_size = DT_INST_PROP_LEN(0, animations);
+
+/**
+ * Pixel configuration.
+ */
+static const struct animation_pixel pixels[] = {
+    ZMK_DT_INST_FOREACH_PROP_ELEM(0, pixels, PHANDLE_TO_PIXEL)};
+
+/**
+ * Size of the pixels array.
+ */
+static const size_t pixels_size = DT_INST_PROP_LEN(0, pixels);
+
+/**
+ * Buffer for RGB values ready to be sent to the drivers.
+ */
+static struct led_rgb px_buffer[DT_INST_PROP_LEN(0, pixels)];
+
+/**
+ * Conditional implementation of zmk_animation_get_pixel_by_key_position
+ * if key-pixels is set.
+ */
+#if DT_INST_NODE_HAS_PROP(0, key_position)
+static const uint8_t pixels_by_key_position[] = DT_INST_PROP(0, key_pixels);
+
+size_t zmk_animation_get_pixel_by_key_position(size_t key_position) {
+    return pixels_by_key_position[key_position];
+}
+#endif
+
+#if defined(CONFIG_ZMK_ANIMATION_PIXEL_DISTANCE) && (CONFIG_ZMK_ANIMATION_PIXEL_DISTANCE == 1)
+
+/**
+ * Lookup table for distance between any two pixels.
+ */
+static uint16_t pixel_distance[DT_INST_PROP_LEN(0, pixels)][DT_INST_PROP_LEN(0, pixels)];
+
+uint16_t zmk_animation_get_pixel_distance(size_t pixel_idx, size_t other_pixel_idx) {
+    return pixel_distance[pixel_idx][other_pixel_idx];
+}
+
+#endif
+
+static void zmk_animation_tick(struct k_work *work) {
+    for (size_t i = 0; i < animations_size; ++i) {
+        animation_on_before_frame(animations[i]);
+    }
+
+    for (size_t i = 0; i < pixels_size; ++i) {
+        struct zmk_color_rgb rgb = {
+            .r = 0,
+            .g = 0,
+            .b = 0,
+        };
+
+        animation_render_pixel(pixels[i].animation, &pixels[i], &rgb);
+
+        zmk_rgb_to_led_rgb(&rgb, &px_buffer[i]);
+    }
+
+    size_t pixels_updated = 0;
+
+    for (size_t i = 0; i < drivers_size; ++i) {
+        led_strip_update_rgb(drivers[i], &px_buffer[pixels_updated], pixels_per_driver[i]);
+
+        pixels_updated += (size_t)pixels_per_driver;
+    }
+
+    for (size_t i = 0; i < animations_size; ++i) {
+        animation_on_after_frame(animations[i]);
+    }
+}
+
+K_WORK_DEFINE(animation_work, zmk_animation_tick);
+
+static void zmk_animation_tick_handler(struct k_timer *timer) { k_work_submit(&animation_work); }
+
+K_TIMER_DEFINE(animation_tick, zmk_animation_tick_handler, NULL);
+
+static int zmk_animation_init(const struct device *dev) {
+#if defined(CONFIG_ZMK_ANIMATION_PIXEL_DISTANCE) && (CONFIG_ZMK_ANIMATION_PIXEL_DISTANCE == 1)
+    // Prefill the pixel distance lookup table
+    for (size_t i = 0; i < pixels_size; ++i) {
+        for (size_t j = 0; j < pixels_size; ++j) {
+            pixel_distance[i][j] = sqrt(pow(pixels[i].position_x - pixels[j].position_x, 2) +
+                                        pow(pixels[i].position_y - pixels[j].position_y, 2));
+        }
+    }
+#endif
+
+    LOG_INF("ZMK Animation Ready");
+
+    k_timer_start(&animation_tick, K_NO_WAIT, K_MSEC(1000 / CONFIG_ZMK_ANIMATION_FPS));
```

Comment:
Do we really need to *always* run a timer for animations? Constantly re-calculating the animations for a solid color backlight would be really inefficient, for example.

Going with my idea above of having animations know about pixels instead of pixels knowing about animations, I was thinking that a more general solution would be something like this:

1. Create a timer exactly like `animation_tick` here, but do not start it right away.
2. Create a pair of `zmk_animation_timer_ref()` and `zmk_animation_timer_unref()` functions. These increment/decrement a ref count, and when the count is > 0, the timer runs. When it is 0, the timer is stopped. This gives us a way for animations to indicate if they want to use the animation timer, so we don't run it when nothing needs to update.
3. Change the animation driver API so that we can do the following:
    1. Activate/deactivate an animation
    2. Tell the animation to update all pixels that it's configured to control
    3. There should also be a way for an animation to request a single animation tick.
4. On system init, activate the default animation.
5. When an animation is activated, it could do the following:
    1. If the animation wants to update at a consistent rate, add a ref to the animation timer.
    2. If the animation wants to update on specific events, e.g. caps lock state or layer state changes, register for those notifications. In its notification handler, it could call the request tick function if the LED state needs to change.
    3. If the animation composes one or more animations, activate those as well.
6. When an animation is deactivated, it would undo everything it did on activation.

Some examples of how that might work:

**Solid color animation**

* Configure with color and optionally a pixel mask to only update a subset of pixels
* On activate, do nothing.
* On tick, set all pixels to the configured color

**Rainbow wave animation**

* Optionally configure with a pixel mask to only update a subset of pixels
* On activate, add ref to animation timer
* On deactivate, unref timer
* On tick, increment animation state and update pixels

**Status indicator animation**

Sets pixels if caps lock, num lock, etc... are enabled.

* Configure with an optional pixel and color for caps lock, num lock, etc... and an optional child animation
* On activate, register for status LED notifications. Forward to child animation if set.
* On caps status change, request a single animation tick.
* On tick, call child animation's tick function if set, then overwrite each status pixel if the corresponding status pixel is enabled.

**Layer highlight animation**

Highlights keys with a function in the current layer by disabling all other LEDs.

* Configure with a single child animation.
* On activate, register for layer change notification. Forward to child animation.
* On layer change, update a mask of which keys have a function in the current layer, then request a single animation tick.
* On tick, call child animation's tick function, then use key mask to overwrite pixels for inactive keys to black.

**Composite animation animation**

Runs multiple animations at the same time. (Each animation would need to be configured to update a subset of pixels, or later animations would just overwrite the output of earlier ones.)

* Configure with one or more child animations.
* On activate, call each child animation's activate function
* On tick, call each child animation's tick function

**Animation selector animation**

Runs a single animation out of a list. Could add key behaviors to select a specific animation or cycle through animations.

* Configure with one or more child animations.
* On activate, activate the first child animation. Register for notifications from animation change behaviors.
* On animation change notification, deactivate current child animation, change selected animation, activate new child animation, and request a single animation tick.
* On tick, call only the selected animation's tick function

Reply:
Thank you for taking a look and the in-depth write up.

### Timer

As for the timer, I haven't played around with that at all yet to be honest. I agree some optimizations could be made there. I'd say we could generally divide animations into two categories: static (only changes on some event) and dynamic (some kind of fluid motion). Dynamic animations triggered by events - think effects - could be considered static when they're idling. So we could try to implement some scheduling to only run at `x` fps when any animation requires it. Otherwise they could just request a new frame to be rendered.

### Activating and deactivating animations

This is something I've been thinking about... taking animation composition into account. I'm not sure if this is something that should be done at the system level or rather having another 'higher-order' animation that toggles everything below it on or off.  
The only advantage of that I can think of right now is when you'd have an animation span multiple zones, and would only like to disable it in one of them. While that sounds like an edge case, I think I actually have an example for this:

On my board I have an indicator LED strip, while it's meant to do several things on its own, you could have it match the keys when idling.  
But whenever it's actually doing something, I wouldn't want the 'shared animation' to stop because it's not used in just that one zone.

Reply:
I think what you listed under "timer" is pretty similar to what I suggested, just that in how I described it, it would be up to the animation to decide how it updates instead of it being something managed by the animation controller. 

Any animation that needs to smoothly animate would ref the timer for as long as it needs it, then unref it whenever it doesn't need to anymore. For basic things that constantly animate, you'd just add a ref when the animation is activated and remove it when deactivated. For an animation that for example makes an effect when you press a key, it could add a ref on every key press and remove it once the animation for that key is over.

Reply:
Interesting... I guess there might still be some merit in having a single timer for actually displaying the frames. Or to be more precise, you'd only want to act on these 'update events' once every 1/30th of a second or something like that rather than a do full update on every single one.  
Like if a single keystroke triggers two animations to update, that will trigger two events at almost the same time, but we don't necessarily want or need to update the LED values twice for that.

But going back to [my other comment](https://github.com/zmkfirmware/zmk/pull/1046#discussion_r767490143), I think I understand it now.  
`render_frame` could be dropped entirely and animations would update their state/buffer internally through events or their own timers. And then it's up to the controller to fetch these values every 1/30th of a second of whatever the setting is - assuming any animation signaled being updated since the last update.

Does that sound reasonable?